### ISSUE TITLE
Font workflow needs some improvement

### ISSUE THREAD

**Issue by QbieShay** [godotengine/godot#24255]
**Godot version:**
11d7738

**Issue description:**
Working with fonts is, currently, rather painful.
I needed to resize a font via code and the only way i found to do so was to use the get("custom_fonts/font") function and the set("custom_fonts/font").
This can either be an issue of documentation, or the structure of the font API itself, since it does not seem that there is a way to set a custom font via code with a method on the control node.
Setting/modifying the current font is rather important since right now the only way to dynamically resize something is to duplicate the font, change the size and set it back.

**Comment by QbieShay** [comment]
Related to #21706

**Comment by razcore-art** [comment]
I agree, it's really painful to work with fonts, probably one of the least fun part of the working with themes/control nodes.

I propose we take out `font size` from the font resource and instead place it on the control nodes, just like other related settings, eg.: `font color` on `Button`. `font size` should be a setting on the control node itself.
There's almost never a use to have a regular font style with a set size and a bold font style with another size setting in say a `RichTextLabel`.

It just doesn't make sense to have to create N amount of font resources for N different font sizes. What other options could we use?

**Comment by christoffersch** [comment]
Definitely agree.
The only way to change the font 'dynamically' is pretty much only via RectSize, which distorts the font.

**Comment by Calinou** [comment]
@christoffersch If you plan to resize a font in a real-time manner, re-rasterizing it every frame isn't an option, so there's going to be bluriness unless you render the font at an higher size and scale it down (with mipmaps enabled, it might look good enough).

Another option is using distance field fonts, but their implementation in Godot is much less complete than DynamicFont. Moreover, these don't look great at small sizes.

**Comment by christoffersch** [comment]
@Calinou I'd have to try scaling it from a higher size and down, thanks.
However, I primarily meant that the only way to change font size in individual control nodes is through RectSize, which is not very optimal :) It would be great to see a font size option for individual nodes.

**Comment by eligt** [comment]
I've made a pull request (https://github.com/godotengine/godot/pull/29762) that would allow specifying a custom font size for each label, it's not perfect as the font is scaled but it's generally good enough if the custom font size is not much smaller or bigger than the original font size.

**Comment by follower** [comment]
### Summary of the comments so far

@QbieShay originally commented:
> I needed to resize a font via code and the only way i found to do so was to use the get("custom_fonts/font") function and the set("custom_fonts/font").
> [...]
> Setting/modifying the current font is rather important since **right now the only way to dynamically resize something is to duplicate the font, change the size and set it back**.

Then @razcore-art commented:
> It just doesn't make sense to have to create N amount of font resources for N different font sizes.

But then @christoffersch commented:
> The only way to change the font 'dynamically' is pretty much only via RectSize, which distorts the font.

Which seems to contradict what @QbieShay said.

### An example of changing the size via GDScript

So, to clarify by way of example: assuming that `$"../Label"` refers to a valid `Label` object (with an attached `DynamicFont` instance), it's possible to modify the attached `DynamicFont` at runtime via GDScript with, e.g.:

```GDscript
$"../Label".get("custom_fonts/font").size = 10 + randi() % 52
```

I'm not 100% sure about when duplication (via GDscript) is necessary to avoid all controls referring to the same resource also having their font size modified but that's certainly possible programmatically. 

### Rasterized font+size characters are cached (as I understand it)

Then @Calinou commented:
> If you plan to resize a font in a real-time manner, re-rasterizing it every frame isn't an option

However, unless one is actually changing the font size 30 times a second, my understanding is that the rasterization of the new size only happens once for each character and the result is cached (I think per font resource/size combination).

Here's some of the relevant Godot engine code:

https://github.com/godotengine/godot/blob/77e894768798c6d6b02759edf5774fb0cd09ac04/scene/resources/dynamic_font.cpp#L710-L730

https://github.com/godotengine/godot/blob/77e894768798c6d6b02759edf5774fb0cd09ac04/scene/resources/dynamic_font.cpp#L45-L62

I agree that the font workflow could stand some improvement but I also think there was/is some confusion about the current state as evidenced in this issue.

**Comment by follower** [comment]
With regard to #29762, here is the start of an example GDScript class to provide "sized label" functionality without modification of the Godot engine (tested in 3.1.1):

```gdscript
# LabelSized.gd

extends Label

tool

class_name LabelSized

export var size: int = 32 setget size_set

func size_set(new_value):

    # TODO: Ensure custom font is set.

    size = new_value
    $".".get("custom_fonts/font").size = size

```

Once you have this script in your project you can add a `LabelSized` node as you would any other node.

You also need to attach a `DynamicFont` (but the font data can just be specified via a path to a `.ttf` file--you don't *actually* need to create a `DynamicFontData` resource within the editor.)

![godot-label-sized-example-screenshot--annotated](https://user-images.githubusercontent.com/189962/62344822-b83db980-b543-11e9-97cd-0cb7b8acb561.png)

Hopefully this helps you out and provides a starting point for a robust/general implementation.

**Comment by KoBeWi** [comment]
Sooo you set font with `add_font_override()` method, which provides autocompletion for available font "slots". Also label/font size problem is solved in 4.0.

I didn't read everything, but this can be closed already, no?

**Comment by Calinou** [comment]
Indeed, changing the font size easily is now resolved in the `master` branch. You can see it in action here: https://youtu.be/3lm-veva15g?t=170
